controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -680,18 +709,57 @@ private void handleStreamMetadataExists(final long timestamp, CompletableFuture<
     }
 
     @Override
-    CompletableFuture<Map<String, Data>> getTxnInEpoch(int epoch) {
+    CompletableFuture<List<Map.Entry<UUID, ActiveTxnRecord>>> getOrderedCommittingTxnInLowestEpoch() {
         synchronized (txnsLock) {
-            Set<String> transactions = epochTxnMap.get(epoch);
-            Map<String, Data> map;
-            if (transactions != null) {
-                map = activeTxns.entrySet().stream().filter(x -> transactions.contains(x.getKey()))
-                        .collect(Collectors.toMap(Map.Entry::getKey, x -> copy(x.getValue())));
-                map = Collections.unmodifiableMap(map);
-            } else {
-                map = Collections.emptyMap();
-            }
-            return CompletableFuture.completedFuture(map);
+            // take smallest epoch and collection transactions from smallest epoch.
+            List<Long> toPurge = new ArrayList<>();
+            Map<UUID, ActiveTxnRecord> committing = new HashMap<>();
+            AtomicInteger smallestEpoch = new AtomicInteger(Integer.MAX_VALUE);
+            transactionCommitOrder
+                    .forEach((order, txId) -> {
+                        String key = txId.toString();
+                        int epoch = RecordHelper.getTransactionEpoch(txId);
+                        ActiveTxnRecord record = activeTxns.containsKey(key) ? ActiveTxnRecord.fromBytes(activeTxns.get(key).getData()) :
+                                ActiveTxnRecord.EMPTY;
+
+                        switch (record.getTxnStatus()) {
+                            case COMMITTING:
+                                if (record.getCommitOrder() == order) {
+                                    // if entry matches record's position then include it
+                                    committing.put(txId, record);
+                                    if (smallestEpoch.get() > epoch) {
+                                        smallestEpoch.set(epoch);
+                                    }
+                                } else {
+                                    toPurge.add(order);
+                                }
+                                break;
+                            case OPEN:  // do nothing
+                                break;
+                            default:
```

Comment:
I believe we can now use exhaustive enums. This way you can enumerate all the cases here and it will be a compiler error if someone adds a new TxnStatus without updating this code.